src/xopen/__init__.py

```diff
@@ -732,10 +733,19 @@ def xopen(
         detected_format = _detect_format_from_content(filename)
 
     if detected_format == "gz":
-        return _open_gz(filename, mode, compresslevel, threads)
+        opened_file = _open_gz(filename, mode, compresslevel, threads)
     elif detected_format == "xz":
-        return _open_xz(filename, mode)
+        opened_file = _open_xz(filename, mode)
     elif detected_format == "bz2":
-        return _open_bz2(filename, mode, threads)
+        opened_file = _open_bz2(filename, mode, threads)
     else:
-        return open(filename, mode)
+        opened_file = open(filename, mode)
+
+    # The "write" method for GzipFile is very costly. Lots of python calls are
+    # made. To a lesser extent this is true for LzmaFile and BZ2File. By
+    # putting a buffer in between, the expensive write method is called much
+    # less. The effect is very noticable when writing small units such as lines
+    # or FASTQ records.
+    if isinstance(opened_file, _compression.BaseStream) and "w" in mode:
```

Comment:
I’m not super happy about using the "private" `_compression` module, but I can’t come up with anything that looks somewhat clean. Any alternative suggestions that are not ugly? It’s not a problem otherwise.

Reply:
We could check for all three types: GzipFile, LzmaFile, Bz2File but that is maybe somewhat more ugly.  It would be more robust though, and since it is a one-time check, there is not a lot of performance overhead. It would also be more transparent and self-explanatory code.
I will implement that.